src/main/java/io/zeebe/kafka/connect/sink/ZeebeSinkFuture.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright © 2019 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.zeebe.kafka.connect.sink;
+
+import io.grpc.StatusRuntimeException;
+import io.zeebe.client.api.command.FinalCommandStep;
+import java.util.concurrent.CompletableFuture;
+
+class ZeebeSinkFuture extends CompletableFuture<Void> {
+
+  private final FinalCommandStep<Void> command;
+
+  ZeebeSinkFuture(final FinalCommandStep<Void> command) {
+    this.command = command;
+  }
+
+  @SuppressWarnings("unchecked")
+  private CompletableFuture<Void> sendCommand() {
+    return (CompletableFuture<Void>) command.send();
+  }
+
+  CompletableFuture<Void> executeAsync() {
+    sendCommand()
+        .whenCompleteAsync(
+            (aVoid, throwable) -> {
+              if (throwable == null) {
+                this.complete(aVoid);
+              } else if (throwable instanceof StatusRuntimeException) {
+                // handle gRPC errors
+                final StatusRuntimeException statusException = (StatusRuntimeException) throwable;
+                switch (statusException.getStatus().getCode()) {
```

Comment:
What do you think about using two enum sets instead of a large switch? We basically have three cases:

```
if (SUCCESS.contains(code)) {
  complete(null);
} else if (RETRIABLES.contains(code)) {
  executeAsync();
} else {
  completeExceptionally(throwable);
}
```

You can define the enum sets as:

```java
private static final Set<Code> SUCCESS_CODES = EnumSet.of(Code.OK, Code.ALREADY_EXISTS);
private static final Set<Code> RETRIABLE_CODES =
    EnumSet.of(
        Code.CANCELLED,
        Code.DEADLINE_EXCEEDED,
        Code.RESOURCE_EXHAUSTED,
        Code.ABORTED,
        Code.UNAVAILABLE,
        Code.DATA_LOSS);
```

These sets can then also be used in tests.

Reply:
Good idea, I will change it